CDCgov/prime-reportstream
# Setup connection to Trusted-Intermediary using OAuth2

This PR setup RESTTransport to connect to Trusted-Intermediary staging server using OAuth2 (two-legged) authentication.

Test Steps:
1. Checkout the PR
2. ./gradlew clean package
3. docker-compose up
4. ./prime multiple-settings set -i ./settings/staging/0149-etor.yml
5. Setup local vault with
    Path for this secret: FLEXION--ETOR-SERVICE-RECEIVER
    {
      "@type": "UserApiKey",
      "apiKey": "Flexion Private Key from KeyBase",
      "user": "flexion"
    }
6. Download the attached test file: etor-patient-bundle-test-json.txt
7. Copy the content of the json file to Postman Body
8. Set header as follow:
    client: flexion.etor-service-sender
    Content-Type: application/fhir+ndjson
    x-functions-key: 09xifjf....444s
9.) POST: http://192.168.1.180:7071/api/reports
![image](https://github.com/CDCgov/prime-reportstream/assets/20068283/15fa553c-fd03-4355-8958-d7ac08dc9f7a)
10.) Check history API:  GET: http://192.168.1.180:7071/api/waters/report/b953c2b0-80ce-4ea9-a25b-067a7e074974/history
![image](https://github.com/CDCgov/prime-reportstream/assets/20068283/884e7483-05aa-415d-9b42-3b450d5ae27e)

Test file:
[etor-patient-bundle-test-json.txt](https://github.com/CDCgov/prime-reportstream/files/12043933/etor-patient-bundle-test-json.txt)

## Changes
- Added Signed JWT generation from Private Key.
- Added authType: two-legged to transport setting.
- Added fakePrivateKey for unit test to cover the new code.

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue
- https://app.zenhub.com/workspaces/engagement-6166edbd409257001e09f1b2/issues/gh/cdcgov/prime-reportstream/9675

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
